### PR TITLE
Commands Update

### PR DESCRIPTION
**Pull request by Absolutionism**
### Description
This PR aims to update commands by adding:

- The ability to provide multiple scripts for `enable`, `disable`, `reload` and `test`
  - Tab completer has been updated to accommodate this
  - Tab completer filters out scripts if their directory has been provided already for `reload`

- Disabling a directory now only changes the name of the directory instead of all scripts within
  - Still unloads all scripts (including scripts within subdirectories)

- Enabling a directory now only changes the name of the directory instead of all scripts within
  - Enables and parses all scripts and subdirectories
  - Scripts or subdirectories that are disabled with the prefix `-` will be ignored, must reload manually or using `all`

- Adds a `lastReloaded` option for `reload`
  - Including lang entries for `There were no scripts previously reloaded.`
  - #7435 

- Public methods for checking if a file is disabled, whether it be the directory it is in or the file itself.

### File Tree "Simulation" For Disabling

<details>
<summary>Before</summary>

```
scripts/
    folder/
        subfolder/
            subfile1
            subfile2
        file1
        file2
```
`sk disable folder`
```
scripts/
    folder/
        subfolder/
            -subfile1
            -subfile2
        -file1
        -file2
```
Though the reasoning behind recursively iterating through all files within a folder is fine. It doesn't contain the ability to only enable that folder, but any subfolders or files that have been explicitly disabled, will be re-enabled.

</details>

<details>
<summary>After</summary>

```
scripts/
    folder/
        subfolder/
            subfile1
            subfile2
        file1
        file2
```
`sk disable folder`
```
scripts/
    -folder/
        subfolder/
            subfile1
            subfile2
        file1
        file2
```
With this, it becomes more efficient by only renaming the folder that was provided rather than for all files. It will also provide the ability of keeping specific files or subfolders disabled.

</details>

---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** none

### CODE REVIEW

File: src/main/resources/lang/english.lang [hunk 1]
```diff
@@ -74,6 +74,7 @@ skript command:
 		x scripts in folder success: <gold>%2$s <lime>script¦¦s¦ in <gold>%1$s<reset>
 		x scripts in folder error: <gold>%2$s <light red>script¦¦s¦ in <gold>%1$s<reset>
 		empty folder: <gold>%s<reset> does not contain any enabled scripts.
+		no previous reload: <light red>There were no scripts previously reloaded.<reset>
```
Comment (Efnilite) [inline src/main/resources/lang/english.lang:77]:
```suggestion
		no previous reload: <light red>No scripts have been previously reloaded.<reset>
```

File: src/main/resources/lang/dutch.lang [hunk 2]
```diff
@@ -69,6 +69,7 @@ skript command:
 		scripts in folder: alle scripts in <gold>%s<reset>
 		x scripts in folder: <gold>%2$s <lime>script¦¦s¦ in <gold>%1$s<reset>
 		empty folder: <gold>%s<reset> bevat geen ingeschakelde scripts.
+		no previous reload: <light red>Er zijn geen scripts eerder opnieuw geladen.<reset>
```
Comment (Efnilite) [inline src/main/resources/lang/dutch.lang:72]:
```suggestion
		no previous reload: <light red>Er zijn geen scripts die eerder opnieuw zijn geladen.<reset>
```

**Review by Efnilite — CHANGES_REQUESTED** [review]

File: src/main/resources/lang/dutch.lang (hunk 2 above)
Comment (TheAbsolutionism) [reply src/main/resources/lang/dutch.lang:72]: Does this mirror your other requested change?

File: src/main/resources/lang/english.lang (hunk 1 above)
Comment (TheAbsolutionism) [reply src/main/resources/lang/english.lang:77]: I want to see what others say

Comment (cheeezburga) [reply src/main/resources/lang/english.lang:77]: Efys is better

File: src/main/resources/lang/dutch.lang (hunk 2 above)
Comment (Efnilite) [reply src/main/resources/lang/dutch.lang:72]: No this just sounds better

File: src/main/resources/lang/english.lang (hunk 1 above)
Comment (Efnilite) [reply src/main/resources/lang/english.lang:77]: This makes the grammar actually correctly use the present perfect

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 3]
```diff
@@ -169,41 +170,55 @@ public boolean onCommand(CommandSender sender, Command command, String label, St
...
-								} else {
-									if (logHandler.numErrors() == 0) {
-										reloaded(sender, logHandler, timingLogHandler, "x scripts in folder success", fileName, scriptInfo.files);
+						scriptFiles = getScriptsFromArgs(sender, args);
+						if (scriptFiles == null || scriptFiles.length == 0)
+							return true;
+						lastReloaded = args;
+					}
+
+					for (File scriptFile : scriptFiles) {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommand.java:189]: I think it's better for behavior if this disables everything and then re-enables everything. If the scripts are interdependent, this kind of load order (unload a, load a, unload b, load b) could cause some weird behavior. You should also make sure there aren't duplicates.

**Review by sovdeeth — CHANGES_REQUESTED** [review]

**Review by sovdeeth — APPROVED** [review]
Looks ok to me
Some concerns about what counts as disabled but that's a wider concern, not for this pr

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 4]
```diff
@@ -497,20 +532,96 @@ private static List<File> getSubFiles(File file) {
...
+	public static boolean scriptIsDisabled(String path) {
+		String[] split = path.split("\\\\");
+		for (String segment : split) {
+			if (segment.startsWith("-"))
+				return true;
+		}
+		return false;
+	}
+
+	public static boolean scriptIsDisabled(File file) {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/SkriptCommand.java:614]: The disabled scripts file filter should already accomplish what this method seeks to

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
Some initial thoughts. I like the idea.

A few things:
- Edge case thought: I have three scripts `chest`, `menu` `chest menu`. If I do `/sk reload chest menu`, what will it determine? That is, how is the conflict handling?
- If I have a path like `scripts/-disabled/test.sk` and I do `/sk reload test.sk`, is `test.sk` allowed to be reloaded, or no since `-disabled` is a disabled directory. I'm not sure what the current behavior is tbh (disabling is rather inconsistent), but I'd like to know what you think should happen.
- How were the translations determined? Did you contact native speakers or use a translation service? If the latter, just to ensure their accuracy, we can always omit the entries in non-English lang files and wait for a native speaker to suggest a translation.

**Comment by Absolutionism** [comment]
> * Edge case thought: I have three scripts `chest`, `menu` `chest menu`. If I do `/sk reload chest menu`, what will it determine? That is, how is the conflict handling?

That is a scenario I actually did not think of, in that instance, it would reload `chest` and `menu` and ultimately make  `chest menu` unable to ever be reloaded. Which is rather a major flaw. The only on the spot potential fix is to make it where the user would need to add `"` to the beginning and end -> `"chest menu"` and I can even have the tab complete be changed to that as well. Which unfortunately would be a breaking change.
Sound good? or any recommendations?

> * If I have a path like `scripts/-disabled/test.sk` and I do `/sk reload test.sk`, is `test.sk` allowed to be reloaded, or no since `-disabled` is a disabled directory.

Assuming the `test.sk` in the `reload` is the one within the disabled folder, it shouldn't reload and also should not be listed in the tab completes for reload.

> * I'm not sure what the current behavior is tbh (disabling is rather inconsistent), but I'd like to know what you think should happen.

The current behavior when disabling a directory, iterates through all files, including files in subdirectories, and disabling them via adding the `-` prefix to the file name but does not  add the prefix to the directory that was provided.
My goal is to only change the file name of the directory that is listed, saves resources as it doesnt have to keep making new files.
It still behaves the same in the sense that all files in the directory, including files in subdirectories, will become disabled.

Let me know your thoughts and anyone  else is welcome to share thoughts and feedback.

**Comment by Absolutionism** [comment]
> That is a scenario I actually did not think of, in that instance, it would reload `chest` and `menu` and ultimately make `chest menu` unable to ever be reloaded. Which is rather a major flaw. The only on the spot potential fix is to make it where the user would need to add `"` to the beginning and end -> `"chest menu"` and I can even have the tab complete be changed to that as well.

An alternative suggested by @Fusezion is instead of the quotation marks `"`, to use commas to separate files i.e `/sk reload chest, menu, chest menu`

**Comment by sovdeeth** [comment]
I like the commas

File: src/main/java/ch/njol/skript/SkriptCommand.java (hunk 4 above)
Comment (Absolutionism) [reply src/main/java/ch/njol/skript/SkriptCommand.java]: The file filter only checks if the file is disabled via if the file itself has the `-` prefix.
So with the change of only changing directory names and not the files within, it would then say that the file is enabled when it's not.

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 5]
```diff
@@ -169,29 +174,65 @@ public boolean onCommand(CommandSender sender, Command command, String label, St
...
+						if (scriptFiles == null || scriptFiles.length == 0)
+							return true;
+					} else {
+						scriptFiles = getScriptsFromArgs(sender, args);
+						if (scriptFiles == null || scriptFiles.length == 0)
+							return true;
+						lastReloaded = args;
+					}
+
+					List<File> scripts = new ArrayList<>();
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommand.java:193]: These could probably by Sets instead of Lists

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 6]
```diff
@@ -169,29 +174,65 @@ public boolean onCommand(CommandSender sender, Command command, String label, St
...
+					for (File scriptFile : scriptFiles) {
 						if (ScriptLoader.getDisabledScriptsFilter().accept(scriptFile)) {
-							info(sender, "reload.script disabled", scriptFile.getName().substring(ScriptLoader.DISABLED_SCRIPT_PREFIX_LENGTH), StringUtils.join(args, " ", 1, args.length));
-							return true;
+							info(sender, "reload.script disabled", scriptFile.getName().substring(ScriptLoader.DISABLED_SCRIPT_PREFIX_LENGTH), scriptFile.getName());
+							continue;
+						}
+						if (!scriptFile.isDirectory()) {
+							Script script = ScriptLoader.getScript(scriptFile);
+							if (!scripts.contains(scriptFile)) {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommand.java:205]: Shouldn't getScript only be called if this condition passes?

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 7]
```diff
@@ -224,111 +265,116 @@ public boolean onCommand(CommandSender sender, Command command, String label, St
...
-						Set<File> scriptFiles;
-						try {
-							scriptFiles = toggleFiles(scriptFile, true);
-						} catch (IOException e) {
-							error(sender, "enable.folder.io error", scriptFile.getName(), ExceptionUtils.toString(e));
-							return true;
-						}
+							final String fileName = scriptFile.getName();
+							info(sender, "enable.single.enabling", fileName);
+							ScriptLoader.loadScripts(scriptFile, logHandler)
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommand.java:288]: should we really be individually enabling every script, with feedback sent for every single one? Shouldn't this batch the enables together?

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 8]
```diff
@@ -224,111 +265,116 @@ public boolean onCommand(CommandSender sender, Command command, String label, St
...
 
-						Set<File> scripts;
-						try {
-							scripts = toggleFiles(scriptFile, false);
-						} catch (IOException e) {
-							error(sender, "disable.folder.io error", scriptFile.getName(), ExceptionUtils.toString(e));
-							return true;
-						}
+							try {
+								toggleFile(scriptFile, false);
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommand.java:357]: Same concern here about doing everything one by one

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 9]
```diff
@@ -224,111 +265,116 @@ public boolean onCommand(CommandSender sender, Command command, String label, St
...
+
+							int totalSubFiles = getSubFiles(scriptFile).size();
+
+							// We're disabling a directory, so we only want to change the name of the directory
+							// After the files within are unloaded
+							FileUtils.move(
+								scriptFile,
+								new File(scriptFile.getParentFile(), ScriptLoader.DISABLED_SCRIPT_PREFIX + scriptFile.getName()),
+								false
+							);
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommand.java:375]: Where did the empty folder error go?

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 10]
```diff
@@ -224,111 +265,116 @@ public boolean onCommand(CommandSender sender, Command command, String label, St
...
-							info(sender, "enable.folder.empty", scriptFile.getName());
-							return true;
-						}
+							Set<File> enabledFiles;
+							try {
+								enabledFiles = toggleFiles(enabledDirectory, true, false);
+							} catch (IOException e) {
+								error(sender, "enable.folder.io error", enabledDirectory.getName(), ExceptionUtils.toString(e));
+								return true;
+							}
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommand.java:313]: Empty folder error?

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 11]
```diff
@@ -398,60 +444,65 @@ public boolean onCommand(CommandSender sender, Command command, String label, St
...
-						scriptFile = getScriptFromArgs(sender, args, TestMode.TEST_DIR.toFile());
-						TestMode.lastTestFile = scriptFile;
+						scriptFiles = getScriptsFromArgs(sender, args, TestMode.TEST_DIR.toFile());
+						TestMode.lastTestFiles = scriptFiles;
 					}
 				}
 
-				if (scriptFile == null || !scriptFile.exists()) {
-					Skript.error(sender, "Test script doesn't exist!");
+				if (scriptFiles == null || scriptFiles.length == 0)
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommand.java:463]: why was the exists() check removed? was it unnecessary? Same with the corresponding error message.

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 12]
```diff
@@ -398,60 +444,65 @@ public boolean onCommand(CommandSender sender, Command command, String label, St
...
-								Skript.exception(e, "Failed to write test results.");
-							}
-						})
-					);
+				for (File scriptFile : scriptFiles) {
+					if (!scriptFile.exists()) {
+						Skript.error("The test script '" + scriptFile.getName() + "' doesn't exist!");
+						continue;
+					}
+					ScriptLoader.loadScripts(scriptFile, errorCounter)
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommand.java:478]: again about doing everything one by one instead of as a batch, especially since here each subsequent script will overwrite the previous results in the result file.

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 13]
```diff
@@ -480,6 +531,45 @@ public boolean onCommand(CommandSender sender, Command command, String label, St
 		return true;
 	}
 
+	private static void reloadScriptFiles(
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommand.java:534]: this isn't actually reloading anything, just loading
Also really really really needs docs with this many parameters

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 14]
```diff
@@ -497,20 +587,128 @@ private static List<File> getSubFiles(File file) {
...
-	private static @Nullable File getScriptFromArgs(CommandSender sender, String[] args) {
-		return getScriptFromArgs(sender, args, Skript.getInstance().getScriptsFolder());
+	private static File @Nullable [] getScriptsFromArgs(CommandSender sender, String[] args) {
+		return getScriptsFromArgs(sender, args, Skript.getInstance().getScriptsFolder());
+	}
+
+	private static File @Nullable [] getScriptsFromArgs(CommandSender sender, String[] args, File directoryFile) {
+		boolean enable = args[0].equals("enable");
+		Map<String, File> scripts = new HashMap<>();
+		for (String currentScript : parseAsCommaSeparatedList(false, true, args)) {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommand.java:597]: currentScript and thisScript are very confusing names and could be improved. For example, currentScript isn't a script, it's a script name from the args. The name should reflect that. Likewise, what does this mean in thisScript?

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 15]
```diff
@@ -497,20 +587,128 @@ private static List<File> getSubFiles(File file) {
...
+		Set<Entry<String, File>> entries = scripts.entrySet();
+		// Now we check to see if any of the files found is already set to be loaded if the directory was also provided
+		for (Entry<String, File> entry : entries) {
+			boolean add = true;
+			for (Entry<String, File> other : entries) {
+				if (entry.getKey().equals(other.getKey()))
+					continue;
+
+				if (entry.getKey().contains(other.getKey()) && entry.getValue().getPath().contains(other.getValue().getPath())) {
+					add = false;
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommand.java:618]: This could be done better with a label and a labelled continue.

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 16]
```diff
@@ -497,20 +587,128 @@ private static List<File> getSubFiles(File file) {
...
+				Skript.error(sender, (directory ? m_invalid_folder : m_invalid_script).toString(currentScript));
+				continue;
+			}
+			scripts.put(currentScript, thisScript);
+		}
+
+		Map<String, File> filteredScripts = new HashMap<>();
+		Set<Entry<String, File>> entries = scripts.entrySet();
+		// Now we check to see if any of the files found is already set to be loaded if the directory was also provided
+		for (Entry<String, File> entry : entries) {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommand.java:611]: I think this whole set of for loops could be replaced with a stream and a reduce() call

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 17]
```diff
@@ -497,20 +587,128 @@ private static List<File> getSubFiles(File file) {
...
+		return isScriptDisabled(file.getPath());
+	}
+
+	/**
+	 * Checks if the provided {@code path} from a file is disabled by checking if the file and the directory parents
+	 * are disabled via the prefix '-'
+	 * @param path
+	 * @return {@code true} if disabled
+	 */
+	public static boolean isScriptDisabled(String path) {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommand.java:649]: This isn't really reliable on its own. You can have a script that's named -test.sk but still be running on the server by renaming it while it's enabled. Maybe not worth addressing here but definitely something to mention in the docs.

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 18]
```diff
@@ -497,20 +587,128 @@ private static List<File> getSubFiles(File file) {
...
+		String[] split = path.split("\\\\");
+		for (String segment : split) {
+			if (segment.startsWith("-"))
+				return true;
+		}
+		return false;
+	}
+
+	public static List<String> parseAsCommaSeparatedList(boolean reverse, boolean excludeEmptyArgs, String ... strings) {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommand.java:658]: docs needed, this goes for all the other helpers too.
Comments within the code explaining what parts do would also be welcome.

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 19]
```diff
@@ -536,26 +734,36 @@ private static File toggleFile(File file, boolean enable) throws IOException {
...
 	}
 
-	private static Set<File> toggleFiles(File folder, boolean enable) throws IOException {
-		FileFilter filter = enable ? ScriptLoader.getDisabledScriptsFilter() : ScriptLoader.getLoadedScriptsFilter();
+	private static Set<File> toggleFiles(File folder, boolean enable, boolean change) throws IOException {
+		return toggleFiles(folder, enable, change, isScriptDisabled(folder));
+	}
 
-		Set<File> changed = new HashSet<>();
+	private static Set<File> toggleFiles(File folder, boolean enable, boolean change, boolean previouslyDisabled) throws IOException {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommand.java:741]: Docs! especially necessary for boolean arguments. enable what? change what?

File: src/main/java/ch/njol/skript/SkriptCommandTabCompleter.java [hunk 20]
```diff
@@ -34,52 +34,77 @@ public List<String> onTabComplete(CommandSender sender, Command command, String
...
-
-						String fileString = file.toString().substring(scriptsPathLength);
-						if (fileString.isEmpty())
-							return;
-
-						if (file.isDirectory()) {
-							fileString = fileString + fs; // Add file separator at the end of directories
-						} else if (file.getParentFile().toPath().toString().equals(scriptsPathString)) {
-							fileString = fileString.substring(1); // Remove file separator from the beginning of files or directories in root only
+			if (args.length == 2 || !args[1].matches("(?i)(all|scripts|aliases|config|lastReloaded)")) {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommandTabCompleter.java:44]: This condition confuses me. What's it meant to do?

File: src/main/java/ch/njol/skript/SkriptCommandTabCompleter.java [hunk 21]
```diff
@@ -34,52 +34,77 @@ public List<String> onTabComplete(CommandSender sender, Command command, String
...
-					});
-			} catch (Exception e) {
-				//noinspection ThrowableNotThrown
-				Skript.exception(e, "An error occurred while trying to update the list of disabled scripts!");
+							// Just in case
+							if (fileString.isEmpty())
+								return;
+
+							// Need to remove files that are already listed or included in a directory
+							if (args[0].matches("(?i)(reload|disable|test)") && args.length > 2 && !separatedArgs.isEmpty()) {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommandTabCompleter.java:86]: basically all of this condition should be done outside the foreach

**Review by sovdeeth — CHANGES_REQUESTED** [review]
